metadata/src/main/java/org/apache/kafka/controller/DelegationTokenControlManager.java

```diff
@@ -0,0 +1,380 @@
+    }
+
+    public ControllerResult<ExpireDelegationTokenResponseData> expireDelegationToken(
+        ControllerRequestContext context,
+        ExpireDelegationTokenRequestData requestData,
+        MetadataVersion metadataVersion
+    ) {
+        long now = time.milliseconds();
+
+        List<ApiMessageAndVersion> records = new ArrayList<>();
+        ExpireDelegationTokenResponseData responseData = new ExpireDelegationTokenResponseData();
+
+        if (secretKeyString == null) {
+            // DelegationTokens are not enabled
+            return ControllerResult.atomicOf(records, responseData.setErrorCode(DELEGATION_TOKEN_AUTH_DISABLED.code()));
+        }
+
+        TokenInformation myTokenInformation = getToken(requestData.hmac());
+
+        if (myTokenInformation == null) {
+            return ControllerResult.atomicOf(records, responseData.setErrorCode(DELEGATION_TOKEN_NOT_FOUND.code()));
+        }
+
+        if (myTokenInformation.maxTimestamp() < now || myTokenInformation.expiryTimestamp() < now) {
+            return ControllerResult.atomicOf(records, responseData.setErrorCode(DELEGATION_TOKEN_EXPIRED.code()));
+        }
+
+        if (!allowedToRenew(myTokenInformation, context.principal())) {
+            return ControllerResult.atomicOf(records, responseData.setErrorCode(DELEGATION_TOKEN_OWNER_MISMATCH.code()));
+        }
+
+        if (requestData.expiryTimePeriodMs() < 0) { // expire immediately
+            responseData
+                .setErrorCode(NONE.code())
+                .setExpiryTimestampMs(requestData.expiryTimePeriodMs());
+            records.add(new ApiMessageAndVersion(new RemoveDelegationTokenRecord().
+                setTokenId(myTokenInformation.tokenId()), (short) 0));
+        } else {
+            long expiryTimestamp = Math.min(myTokenInformation.maxTimestamp(),
+                now + requestData.expiryTimePeriodMs());
+
+            responseData
+                .setErrorCode(NONE.code())
+                .setExpiryTimestampMs(expiryTimestamp);
+
+            myTokenInformation.setExpiryTimestamp(expiryTimestamp);
+
+            DelegationTokenData newDelegationTokenData = new DelegationTokenData(myTokenInformation);
+            records.add(new ApiMessageAndVersion(newDelegationTokenData.toRecord(), (short) 0));
+        }
+
+        return ControllerResult.atomicOf(records, responseData);
+    }
+
+    // Periodic call to remove expired DelegationTokens
+    public List<ApiMessageAndVersion> expireDelegationTokens() {
+        long now = time.milliseconds();
+        List<ApiMessageAndVersion> records = new ArrayList<ApiMessageAndVersion>();
+
+        for (TokenInformation oldTokenInformation: tokenCache.tokens()) {
+            if ((oldTokenInformation.maxTimestamp() < now) ||
+                (oldTokenInformation.expiryTimestamp() < now)) {
+                System.out.println("Token: " + oldTokenInformation.tokenId() + " is expired");
```

Comment:
Should use Logger instead of System.out here. It might be useful to include whether the expiration was due to max timestamp or the expiry timestamp.

Reply:
Yes, this was a debug statement that I missed. I updated to use the exact same log line from the Zk expireTokens() now in DelegationTokenManagerZk.scala.